File: pset-1/Credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>
bool validar_credito(long long numero_credito);
//vai checar se o cartão eh verdadeiro se ele esta em 13,15 ou 15 casas e checar em checksum
bool checksum(long long divi);
//vai checar se a soma esta correta de ex:4012 - 4*2+1*2= 10+0+2= 12/10== 0
int enc_tam(long long y);
//vai identificar o tamanho de casas do cartao
void print_bandeira(long long divi);

//american express : 15 digitos,
int main(void)
{
    long long numero_credito;
    do
    {
        numero_credito = get_long_long("Número do cartão: ");
//vai pegar o nuhmero do cartão
    }
    while (numero_credito < 0);
    if (validar_credito(numero_credito) == true)
    {
        print_bandeira(numero_credito);
    }
    else
    {
        printf("INVALID\n");
    }
    return 0;
//vai verificar o cartão e se não passar retorna INVALID
}

bool validar_credito(long long numero_credito)
{
    int tam = enc_tam(numero_credito);
    return (tam == 13 || tam == 15 || tam == 16) && checksum(numero_credito);
//checa o tamanho do cartão usando o enc_tam e verificar se a conta é correspode ao já falado no pset-1
}

int enc_tam(long long y)
{
    int tam;
    for (tam = 0; y != 0; y /= 10, tam++);
    return tam;
//conta a quantidade de casa em credito_cartão e retorna a quantidade
}

bool checksum(long long divi)
{
    int soma = 0, i, digito;
    for (i = 0; divi != 0; i++, divi /= 10)
//vai fazer um laço para contar cada casa em numero_credito
    {
        if (i % 2 == 0)
        {
            soma += divi % 10;
        }
//se for uma casa par ela soma no montante normalmente
        else
        {
            digito = 2 * (divi % 10);
            soma += (digito / 10) + (digito % 10);
//aqui ela pega os impares e multiplica por 2 e soma no montante
        }
    }
    return (soma % 10) == 0;
//nesta parte como eu fiz ela podendo retornar ou true ou false
}

void print_bandeira(long long divi)
{
    if ((divi >= 34e13 && divi < 35e13) || (divi >= 37e13 && divi < 38e13))
    {
        printf("AMEX\n");
    }
    else if (divi >= 51e14 && divi < 56e14)
    {
        printf("MASTERCARD\n");
    }
    else if ((divi >= 4e12 && divi < 5e12) || (divi >= 4e15 && divi < 5e15))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }

//o 34e13 ou 51e14 é uma notacao cientifica de 34x10(elecado a)13
//print_bandeira retorna para qual cartão é cada endereco do cartão e tambem checa se ele é válido
}

File: pset-6/credit.py
# TODO
# vai fazer conta do credit em phython
def checksum(divi):
    i = 0
    soma = 0
    # o loop vai continuar até que divi não seja iqual a zero
    while divi != 0:
        if i % 2 == 0:
            # vai fazer a primeria parte da soma
            soma += divi % 10
        else:
            # vai fazer a segunda parte da soma
            digito = 2 * (divi % 10)
            soma += (digito // 10) + (digito % 10)
        # vai passar por todos as casas
        i += 1
        divi //= 10
    return soma % 10 == 0


# ja vai medir o tamanho e validar o tamanho da quantidade
# de casas do cartão
def validar_credito(numero_credito):
    tam = len(str(numero_credito))
    return (tam == 13 or tam == 15 or tam == 16) and checksum(numero_credito)


# vai determinar qual a bandeira do cartão
def print_bandeira(divi):
    # bandeira AMEX
    if (divi >= 34e13 and divi < 35e13) or (divi >= 37e13 and divi < 38e13):
        print("AMEX\n")
    # bandeira MASTERCARD
    elif (divi >= 51e14 and divi < 56e14):
        print("MASTERCARD\n")
    # bandeira VISA
    elif (divi >= 4e12 and divi < 5e12) or (divi >= 4e15 and divi < 5e15):
        print("VISA\n")
    # caso não seja nenhuma delas irá sair como inválida
    else:
        print("INVALID\n")


# tratamento de excessão
try:
    numero_credito = None
    while True:
        # pega o número do cartão
        numero_credito = int(input("Número do cartão: "))
        # caso seja maior que zero ela sai do while
        if numero_credito >= 0:
            break
    # aqui vai validar e caso tenha uma banderia correspondente
    # ira mostrar o nome dela
    if validar_credito(numero_credito):
        print_bandeira(numero_credito)
    # caso não tenha correspondente irá retornar inválida
    else:
        printf("INVALID\n")
# caso o código não
except Exception as e:
    print("INVALID")
